### ISSUE TITLE
Completion backend hangs when using Corfu as completion frontend

### ISSUE THREAD

**Issue by drcxd** [emacs-lsp/lsp-mode#3555]
### Thank you for the bug report

- [X] I am using the latest version of `lsp-mode` related packages.
- [X] I checked [FAQ](https://emacs-lsp.github.io/lsp-mode/page/faq/) and [Troubleshooting](https://emacs-lsp.github.io/lsp-mode/page/troubleshooting/) sections
- [ ] You may also try reproduce the issue using clean environment using the following command: `M-x lsp-start-plain`

### Bug description

Sometimes, lsp completion backend hangs when using Corfu as the completion front end.

Here is the profile:
```
         731  80% - corfu--post-command
         730  80%  - corfu--update
         726  80%   - corfu--update-candidates
         723  79%    - corfu--recompute-candidates
         723  79%     - corfu--all-completions
         723  79%      - completion-all-completions
         723  79%       - apply
         723  79%        - #<compiled -0x187e8d863e1757aa>
         723  79%         - completion--nth-completion
         722  79%          - completion--some
         722  79%           - #<compiled 0xf84d7d713b6baaf>
         722  79%            - orderless-all-completions
         722  79%             - orderless-filter
         722  79%              - #<compiled -0xd4df90488ccba75>
         722  79%               - #<compiled 0x9ea0f36097de3b9>
         722  79%                - lsp-request-while-no-input
         722  79%                 - sit-for
          70   7%                  + read-event
           5   0%                  + redisplay
           1   0%            completion--styles
           3   0%      redisplay
           3   0%     redisplay
           1   0%   - corfu--candidates-popup
           1   0%      corfu--popup-show
           1   0%  - corfu-quit
           1   0%   + completion-in-region-mode
         116  12% + ...
          29   3% + redisplay_internal (C function)
          17   1% - timer-event-handler
          17   1%  - apply
          17   1%   - corfu--auto-complete
          17   1%    - corfu--update
           5   0%     + corfu--update-candidates
           5   0%     + corfu-quit
           4   0%     + corfu--candidates-popup
           3   0%       redisplay
          12   1% - command-execute
          12   1%  - funcall-interactively
           8   0%   - corfu-insert
           8   0%    - corfu--insert
           8   0%     - corfu--done
           8   0%      + #<compiled -0x7aa2ac626036bae>
           4   0%   - self-insert-command
           2   0%    - tree-sitter--after-change
           2   0%       tree-sitter--do-parse
           1   0%   evil--jump-hook
```

More details in this issue: https://github.com/minad/corfu/issues/188

### Steps to reproduce

It is quite troublesome to reproduce this issue on any computer, I am not sure could you find the problem through the profiler report. If you really need to reproduce this, please let me know and I will prepare the project.

### Expected behavior

When Corfu completion is triggered, it does not hang that long. I saw that `lsp-response-timeout` is defined as `10`.

### Which Language Server did you use?

C++ with Clangd

### OS

Windows

**Comment by c-alpha** [comment]
Could this be another symptom of #3607?

**Comment by justinbarclay** [comment]
Hey, I am running into this issue, Corfu is hanging when using LSP when writing JSX in JavaScript and Typescript. I've noticed the most egregious behaviour when I start to call a component, I.E. a `<` followed by any character. The pause isn't guaranteed and happens semi-randomly, so it's hard for me to create a test repo. I also noticed it happening a lot when defining a self-closing element or component `/>`. Typing `/` was fine, but when I tried to type `>`, it would freeze for multiple seconds.

**Comment by c-alpha** [comment]
As @minad points out, there is definitely something to be fixed on the
LSP Mode side.

There may be a second issue lurking though. I've had random lock-ups
with Emacs 28 and 29, independent of corfu and/or lsp-mode.
Single-stepping in the debugger seemed to hint to an infinite loop
somewhere in the code that tries to compute the font settings.

Due to other problems ("symbol's definition is void"), I had to do a
`(byte-recompile-directory package-user-dir nil 'force)`, and have never
had another lock-up since. Also, a loaddefs related patch has been
pushed to Emacs 29 recently. Thus, FWIW, you might want to try the
`byte-recompile-directory` thing and see if that changes anything for
you?

Just my two cents anyway.

**Comment by wyuenho** [comment]
Does anyone have a backtrace? M-x toggle-debug-on-error and try to reproduce it

**Comment by wyuenho** [comment]
> I had the same problem and https://github.com/minad/corfu/issues/188#issuecomment-1148658471 fixed it for me:
> 
> ```elisp
> (advice-add #'lsp-completion-at-point :around #'cape-wrap-noninterruptible)
> ```

Were you trying to complete Typescript with ts-ls too?

**Comment by martenlienen** [comment]
No, it was [python-lsp-server](https://github.com/python-lsp/python-lsp-server).
